text/0004-recoverable-transaction-history.md

```diff
@@ -0,0 +1,347 @@
+an arbitrary one). It also includes a count of the number of recipients, and the total outlay and fee paid.
+
+If there is more than one recipient, only the address hash of one of them can be stored in the space available.
+
+The destination memo does not have a MAC, however it can be validated by confirming that the associated TxOut went to the
+change subaddress for this account.
+
+To prevent a malicious party from confusing the sender, a client should ignore this memo unless it is on a TxOut
+that went to the change subaddress. 
+
+## Change subaddress
+
+The "change subaddress" is a new standard subaddress, like the default subaddress. It should be used when sending change outputs.
+It is specified to be subaddress index 1.
+
+The change subaddress should be kept secret, so that only authentic 0x0200 Destination memos will pass validation.
+
+To fully support recoverable transaction history, clients should use 0x0100 or 0x0101 Authenticated sender memos with all
+outputs that they send to another party, and 0x0200 Destination memos on all change outputs, which are all sent to the change
+subaddress. Change outputs should be sent *even* in the case that the change value is zero, in order that there will be a destination memo.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+A 16-byte *short address hash* is specified:
+
+```
+address_hash = public_address.digest32(“mc-address-hash”)[0..16]
+```
+
+Here `digest32` refers to the merlin-based hash computed using `mc-crypto-digestible` crate.
+
+The *change subaddress* is specified as the subaddress of index 1. It should not
+be used except for change outputs, and this subaddress should not be shared with third parties.
+
+All transactions should have exactly *one* change output, even if the change value is zero.
+
+Three new memo types are specified:
+
+| Memo type bytes | Name                                              |
+| -----------     | -----------                                       |
+| 0x0100          | Authenticated Sender Memo                         |
+| 0x0101          | Authenticated Sender With Payment Request Id Memo |
+| 0x0200          | Destination Memo                                  |
+
+## 0x0100 Authenticated Sender Memo
+
+The 44-byte memo data is laid out as follows:
+
+| Byte range | Item |
+| ---------- | ---- |
+| 0 - 16     | Sender's Address Hash |
+| 16 - 28    | Unused bytes          |
+| 28 - 44    | HMAC                  |
+
+The HMAC is computed using HMAC-SHA512. 
+(HMAC refers to [RFC 2104 HMAC](https://datatracker.ietf.org/doc/html/rfc2104).)
+(SHA512 refers to the 512 bit variant of [SHA-2](https://en.wikipedia.org/wiki/SHA-2).)
+
+The key to this HMAC is the 32-byte shared secret computed from Ristretto key exchange
+using the Sender's spend key, and the Recipient's view key. (For the sender, we multiply
```

Comment:
Doesn't this mean the MAC key isn't uniform? Do we need to run this through an HKDF first?

Reply:
(marking as resolved b/c we're using ristretto kex details elsewhere in similar situations, and also HMAC is going to run the key through SHA512 straight away anyways, HKDF just has an extra SHA512 round)